pkg/networkservice/common/refresh/client.go

```diff
@@ -24,79 +24,92 @@ import (
 
 	"github.com/golang/protobuf/ptypes"
 	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
 	"github.com/pkg/errors"
 	"google.golang.org/grpc"
 
-	"github.com/networkservicemesh/api/pkg/api/networkservice"
-
 	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
 	"github.com/networkservicemesh/sdk/pkg/networkservice/core/trace"
 	"github.com/networkservicemesh/sdk/pkg/tools/extend"
 	"github.com/networkservicemesh/sdk/pkg/tools/serialize"
 )
 
+type refreshCanceller struct {
+	timer      *time.Timer
+	cancelFunc func()
+}
+
 type refreshClient struct {
-	connectionTimers map[string]*time.Timer
-	connections      map[string]*networkservice.Connection
-	executor         serialize.Executor
+	cancellers map[string]*refreshCanceller
+	executor   serialize.Executor
 }
 
 // NewClient - creates new NetworkServiceClient chain element for refreshing connections before they timeout at the
 // endpoint
 func NewClient() networkservice.NetworkServiceClient {
 	rv := &refreshClient{
-		connectionTimers: make(map[string]*time.Timer),
-		connections:      make(map[string]*networkservice.Connection),
-		executor:         serialize.NewExecutor(),
+		cancellers: map[string]*refreshCanceller{},
+		executor:   serialize.Executor{},
 	}
 	return rv
 }
 
 func (t *refreshClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
-	rv, err := next.Server(ctx).Request(ctx, request)
+	rv, err := next.Client(ctx).Request(ctx, request)
 	if err != nil {
 		return nil, errors.Wrap(err, "Error calling next")
 	}
 	// Clone the request
 	req := request.Clone()
 	// Set its connection to the returned connection we received
 	req.Connection = rv
-	// Setup the timer with the req containing the returned connection
-	ct, err := t.createTimer(ctx, req, opts...)
+
+	expireTime, err := ptypes.Timestamp(request.GetConnection().GetPath().GetPathSegments()[request.GetConnection().GetPath().GetIndex()].GetExpires())
```

Comment:
@edwarnicke What do you think about having minimal granularity for token expire time, like 5 seconds.
So we could have just one timer to check expires for every 5 seconds for all connections, in this case it won't be required to create times and could be easy doable with executor AsyncExec and sleep.